theano/tensor/nnet/abstract_conv.py

```diff
@@ -1505,7 +1535,13 @@ def conv(self, img, kern, mode="valid", dilation=1):
                                                                  self.convdim))
 
         out_shape = get_conv_output_shape(img.shape, kern.shape,
-                                          mode, [1] * self.convdim, dilation)
+                                          mode, [1] * self.convdim, dilation, unshared)
+
+        if unshared is True:
+            if (kern.shape[2], kern.shape[3]) != (out_shape[2], out_shape[3]):
+                raise ValueError('Kernel shape ({},{}) does not match '
+                                 'output size ({},{})'.format(kern.shape[2], kern.shape[3],
+                                                              out_shape[2], out_shape[3]))
 
         out = np.zeros(out_shape, dtype=img.dtype)
         dil_kern_shp = kern.shape[:-self.convdim] + tuple(
```

Comment:
A few lines below, I think you might need to change this assignment of `dilated_kern`:
```python
dilated_kern[(slice(None), slice(None)) +
             tuple(slice(None, None, dilation[i]) for i in range(self.convdim))
             ] = kern
```
because it doesn't include the row/column dimensions in an unshared convolution.